File: src/C3.Infrastructure/Diagnostics/CrashReportWriter.vb
Imports System.IO
Imports System.Text

Namespace Diagnostics

Public NotInheritable Class CrashReportContext

    Public Property ProductVersion As String
    Public Property ReleaseStage As String
    Public Property BuildLane As String
    Public Property OperatingSystem As String
    Public Property ClrVersion As String
    Public Property ProcessBitness As String
    Public Property CataloguePath As String
    Public Property LastAction As String

End Class

Public NotInheritable Class CrashReportWriter

    Private Sub New()
    End Sub

    Public Shared Function TryWrite(exception As Exception, context As CrashReportContext) As String
        If exception Is Nothing Then
            Return Nothing
        End If

        Try
            Dim localData As String = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)
            Dim reportDirectory As String = Path.Combine(localData, "C3", "CrashReports")
            Directory.CreateDirectory(reportDirectory)

            Dim fileName As String = "C3-error-" & DateTime.Now.ToString("yyyyMMdd-HHmmss-fff") & ".log"
            Dim reportPath As String = Path.Combine(reportDirectory, fileName)
            File.WriteAllText(reportPath, BuildReport(exception, context), New UTF8Encoding(False))
            Return reportPath
        Catch
            Return Nothing
        End Try
    End Function

    Private Shared Function BuildReport(exception As Exception, context As CrashReportContext) As String
        If context Is Nothing Then
            context = New CrashReportContext() With {.LastAction = BufferedLogger.LastAction}
        End If

        Dim report As New StringBuilder()
        report.AppendLine("C3 crash report")
        report.AppendLine("===============")
        report.AppendLine("Created (UTC): " & DateTime.UtcNow.ToString("O"))
        report.AppendLine("Product version: " & If(context.ProductVersion, "(unknown)") & " " & If(context.ReleaseStage, String.Empty))
        report.AppendLine("Build lane: " & If(context.BuildLane, "(unknown)"))
        report.AppendLine("Operating system: " & If(context.OperatingSystem, "(unknown)"))
        report.AppendLine("CLR version: " & If(context.ClrVersion, "(unknown)"))
        report.AppendLine("Process bitness: " & If(context.ProcessBitness, "(unknown)"))

        If context IsNot Nothing Then
            report.AppendLine("Catalogue path: " & If(context.CataloguePath, "(new catalogue)"))
            report.AppendLine("Last action: " & If(context.LastAction, "(unknown)"))
        Else
            report.AppendLine("Catalogue path: (unknown)")
            report.AppendLine("Last action: " & BufferedLogger.LastAction)
        End If

        report.AppendLine()
        report.AppendLine("Exception")
        report.AppendLine("---------")
        report.AppendLine(exception.ToString())
        report.AppendLine()
        report.AppendLine("Recent log")
        report.AppendLine("----------")
        For Each entry As String In BufferedLogger.Tail()
            report.AppendLine(entry)
        Next

        Return report.ToString()
    End Function

End Class

End Namespace

File: src/C3.Infrastructure/Diagnostics/CrashReportWriter.cs
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.IO;
using System.Text;

namespace C3.Infrastructure.Diagnostics
{
	public sealed class CrashReportWriter
	{
		private CrashReportWriter()
		{
		}

		public static string TryWrite(Exception exception, CrashReportContext context)
		{
			if (exception == null)
			{
				return null;
			}
			try
			{
				string text = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "C3", "CrashReports");
				Directory.CreateDirectory(text);
				string path = "C3-error-" + DateTime.Now.ToString("yyyyMMdd-HHmmss-fff") + ".log";
				string text2 = Path.Combine(text, path);
				File.WriteAllText(text2, BuildReport(exception, context), new UTF8Encoding(false));
				return text2;
			}
			catch (Exception projectError)
			{
				ProjectData.SetProjectError(projectError);
				string result = null;
				ProjectData.ClearProjectError();
				return result;
			}
		}

		private static string BuildReport(Exception exception, CrashReportContext context)
		{
			if (context == null)
			{
				context = new CrashReportContext
				{
					LastAction = BufferedLogger.LastAction
				};
			}
			StringBuilder stringBuilder = new StringBuilder();
			stringBuilder.AppendLine("C3 crash report");
			stringBuilder.AppendLine("===============");
			stringBuilder.AppendLine("Created (UTC): " + DateTime.UtcNow.ToString("O"));
			stringBuilder.AppendLine("Product version: " + (context.ProductVersion ?? "(unknown)") + " " + (context.ReleaseStage ?? string.Empty));
			stringBuilder.AppendLine("Build lane: " + (context.BuildLane ?? "(unknown)"));
			stringBuilder.AppendLine("Operating system: " + (context.OperatingSystem ?? "(unknown)"));
			stringBuilder.AppendLine("CLR version: " + (context.ClrVersion ?? "(unknown)"));
			stringBuilder.AppendLine("Process bitness: " + (context.ProcessBitness ?? "(unknown)"));
			if (context != null)
			{
				stringBuilder.AppendLine("Catalogue path: " + (context.CataloguePath ?? "(new catalogue)"));
				stringBuilder.AppendLine("Last action: " + (context.LastAction ?? "(unknown)"));
			}
			else
			{
				stringBuilder.AppendLine("Catalogue path: (unknown)");
				stringBuilder.AppendLine("Last action: " + BufferedLogger.LastAction);
			}
			stringBuilder.AppendLine();
			stringBuilder.AppendLine("Exception");
			stringBuilder.AppendLine("---------");
			stringBuilder.AppendLine(exception.ToString());
			stringBuilder.AppendLine();
			stringBuilder.AppendLine("Recent log");
			stringBuilder.AppendLine("----------");
			foreach (string item in BufferedLogger.Tail())
			{
				stringBuilder.AppendLine(item);
			}
			return stringBuilder.ToString();
		}
	}
}
